### DISCUSSION TITLE
Can you refer to IndirectBuffer and only pass 4 dynamic data of Uint type when writing passEncoder, or even only pass 1 Uint data, which can reduce the waste of video memory.

### DISCUSSION THREAD

**Discussion by GoodLuckIce** [gpuweb/gpuweb#4347] [Q&A] [answered]
Because setBindGroup's dynamicOffsets are limited by minUniformBufferOffsetAlignment, we cannot pass data smaller than 255 bytes.
Can you refer to IndirectBuffer and only pass 4 dynamic data of Uint type when writing passEncoder, or even only pass 1 Uint data, which can reduce the waste of video memory.
In actual business, I need additional data to be used in conjunction with Draw to tell the shader how to parse other Buffer data to correctly handle the rendering logic.

**Comment by GoodLuckIce** [comment]
There are currently a large number of meshes that need to be rendered. Use "Compute Shader" to perform frustum culling, generate an IndirectBuffer containing multiple mesh information, and then call drawIndirect to draw.
However, the IndirectBuffer only contains information about the number of grid vertices and does not have the offset address of the grid vertex data location, resulting in the inability to draw normally.
I hope to modify drawIndirect so that it can pass the offset address of mesh data to meet the above usage scenario.

**Reply by Kangz** [reply, marked as answer]
There seems to be two things you are asking for: the first one is an ability to pass very small "uniform" data in a simpler and more efficient way than uniform buffer dynamic offset. Push constants (#75) could be a good way to do this if/when they are added to WebGPU.

For the second one, would you be able to use push constants to pass the data about which mesh is used or is that only known on the GPU? If it is only known on the GPU you could abuse "first instance" to store that data and then get the `@builtin(instance_index)` in the shader to use as an index in another storage buffer that contains more data. (of course it would require the `indirect-first-instance` feature)

**Reply by JMS55** [reply]
> However, the IndirectBuffer only contains information about the number of grid vertices and does not have the offset address of the grid vertex data location, resulting in the inability to draw normally.

What you can do is upload a storage buffer containing a map from vertex number -> vertex location, and then use @builtin(vertex_index) in your vertex shader + the storage buffer to get the actual vertex location.
